### PR TITLE
merge in new scikit changes

### PR DESCRIPTION
**Pull request by dmaticzka**

### CODE REVIEW

File: mean_shift_sklearn.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+- read file input.bed and write new peaks/centroids into output.bed
+- mean_shift.py input.bed 200 -o data_output.bed -g output-graph.png
+
+"""
+
+
+class Mean_Shift():
+
+    """
+    calcualtes the mean shift
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: better: applies mean shift algorithm to an array of integers

File: mean_shift_sklearn.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            plt.plot(X, X, col + '.', cluster_center[0],
+                     cluster_center[0], 'o',
+                     markerfacecolor=col,
+                     markeredgecolor='k', markersize=14)
+        plt.title('Estimated number of clusters: %d' % n_clusters_)
+        # plt.show()
+        plt.savefig(graph)
+
+    def output_file(self, centroids, output):
+
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: add documentation

File: mean_shift_sklearn.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    >>> a = Mean_Shift(3)
+    >>> x = np.array([[1, 2, 3, 4, 5, 9, 10, 11, 12, 13, 20, 21, 22, 23, 24]])
+    >>> a.peak(x)
+    (array([[ 3.],
+           [11.],
+           [22.]]), 3)
+    """
+    def __init__(self, bandwidth):
+        self.bandwidth = bandwidth
+        self.labels = []
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: what is labels?

Comment (dmaticzka) [reply mean_shift_sklearn.py]: add line comment describing this

Comment (dmaticzka) [reply mean_shift_sklearn.py]: remove if not needed

File: mean_shift_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    def plot_graph(self, cluster_centers, n_clusters_, X, graph):
+        # #############################################################################
+        # Plot result
+        plt.figure(1)
+        plt.clf()
+        colors = cycle('bgrcmykbgrcmykbgrcmykbgrcmyk')
+
+        for k, col in zip(range(n_clusters_), colors):
+            self.labels == k
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: this should not be a class variable

File: mean_shift_sklearn.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    >>> a.peak(x)
+    (array([[ 3.],
+           [11.],
+           [22.]]), 3)
+    """
+    def __init__(self, bandwidth):
+        self.bandwidth = bandwidth
+        self.labels = []
+
+    def peak(self, dt):
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: dt is not a good variable name, it does not tell us what is in it

Comment (dmaticzka) [reply mean_shift_sklearn.py]: add to function documentation: what kind of object is this? how to get it?

File: mean_shift_sklearn.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        # #############################################################################
+        # Compute clustering with MeanShift
+        ms = MeanShift(bandwidth=self.bandwidth, bin_seeding=True)
+        ms.fit(dt.transpose())
+        self.labels = ms.labels_
+        cluster_centers = ms.cluster_centers_
+        labels_unique = np.unique(self.labels)
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: directly use ms.labels_ in function

File: mean_shift_sklearn.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        n_clusters_ = len(labels_unique)
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: don't go via labels, you can directly count the cluster centers

File: mean_shift_sklearn.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        return cluster_centers, n_clusters_
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: only return array, number of clusters is simple to retrieve from that

File: mean_shift_sklearn.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    def plot_graph(self, cluster_centers, n_clusters_, X, graph):
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: same here, generate n_clusters from array

File: mean_shift_sklearn.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                                                  name, score, strand))
+                wr.write(cl + '\n')
+
+
+def main():
+    parser = argparse.ArgumentParser(description=information,
+                                     formatter_class=argparse.
+                                     RawDescriptionHelpFormatter)
+    parser.add_argument("dataFile", type=str, help='Enter the data file name')
+    parser.add_argument("Bandwidth", type=int, help='Enter the bandwidth')
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: lowecase please

File: mean_shift_sklearn.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    parser.add_argument("-o", "--output", default='output.bed',
+                        help='Enter the output data file name')
+    parser.add_argument("-g", "--graph", default="graph.png",
+                        help='Enter the output graph file name')
+
+    Pars = parser.parse_args()
```
Comment (dmaticzka) [inline mean_shift_sklearn.py]: lowercase

Comment (dmaticzka) [reply mean_shift_sklearn.py]: usually this variable is simply calld 'args'

**Review by dmaticzka — COMMENTED** [review]
nothing major :)

File: mean_shift_sklearn.py (hunk 1 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done

File: mean_shift_sklearn.py (hunk 3 above)
Comment (fejjaz) [reply mean_shift_sklearn.py]: done!!!

Comment (fayyazh) [reply mean_shift_sklearn.py]: removed label !!!

File: mean_shift_sklearn.py (hunk 4 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: removed !!!

File: mean_shift_sklearn.py (hunk 7 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 8 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 9 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 10 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done  !!!

File: mean_shift_sklearn.py (hunk 11 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 5 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 2 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: done !!!

File: mean_shift_sklearn.py (hunk 6 above)
Comment (fayyazh) [reply mean_shift_sklearn.py]: removed it !!!

**Comment by dmaticzka** [comment]
good work, thank you!
